liron's `ButteredBread` fixes

lirona opened the pull request:
for #119 i've implemented the simple solution but you could also just make sure you've iterated over all users and that way you don't need to pass an array of users to update and don't need to keep scaling factor per user in the contract or to sync voting weights.

for #117 i've just added a warning, as decided by @RonTuretzky

lirona committed:
add amount 0 checks (issue #114)

lirona committed:
check transfer return value (issue #115)

lirona committed:
move transfer to the end of _deposit (issue #116)

also moved `_syncDelegation` just for consistency

lirona committed:
fix natspec for balanceOfLP (issue #118)

lirona committed:
_syncVotingWeight in deposit (issue #119)

lirona committed:
add warning to modifyAllowList (issue #117)

bagelface commented on src/ButteredBread.sol:
```diff
@@ -75,6 +75,7 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
      * @param _amount Value of LP token
      */
     function deposit(address _lp, uint256 _amount) external onlyAllowed(_lp) {
+        require(_amount != 0, "Amount must be greater than 0");
```
prefer to use "custom errors" approach for assertions.

bagelface commented on src/ButteredBread.sol:
```diff
@@ -84,12 +85,14 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
      * @param _amount Value of LP token
      */
     function withdraw(address _lp, uint256 _amount) external onlyAllowed(_lp) {
+        require(_amount != 0, "Amount must be greater than 0");
```
same here comment about "custom errors" here

bagelface commented on src/ButteredBread.sol:
```diff
@@ -155,7 +158,7 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
 
         _burn(_account, _amount * scalingFactors[_lp] / FIXED_POINT_PERCENT);
-        IERC20(_lp).transfer(_account, _amount);
+        require(IERC20(_lp).transfer(_account, _amount), "Transfer failed");
```
prefer custom errors and something like:
```
bool success = IERC20(_lp).transfer(_account, _amount);
if (!success) revert TransferFailed();

```

bagelface commented on src/ButteredBread.sol:
```diff
@@ -123,25 +126,25 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
+        _accountToLPData[_account][_lp].balance += _amount;
+        
+        _mint(_account, _amount * scalingFactors[_lp] / FIXED_POINT_PERCENT);
+        require(IERC20(_lp).transferFrom(_account, address(this), _amount), "Transfer failed");
```
custom error pattern like:
```
bool success = IERC20(_lp).transferFrom(_account, address(this), _amount)
if (!success) revert TransferFailed();
```
for better readability

bagelface commented on src/ButteredBread.sol:
```diff
@@ -123,25 +126,25 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
-        _mint(_account, _amount * currentScalingFactor / FIXED_POINT_PERCENT);
         _syncDelegation(_account);
 
+        /// @dev ensure proper accounting in case of admin error in `modifyScalingFactor` where not all holders are updated
```
i think the comment isn't necessary here. it can potentially be moved to the natspec function comment

lirona replied:
that was the way it was done by you in the `withdraw` function.. do you want me to move that from there as well?

bagelface replied:
oh i didnt notice. Yea up to you if you want to change it in withdraw, I'll disregard this in the review.

lirona replied:
actually would be good to add a comment in the public functions' natspec, i added it in [this commit](https://github.com/BreadchainCoop/breadchain/pull/120/commits/4e801486fdd8b28ae982455fdef20d9a0b815436), wdyt?

bagelface reviewed (changes requested):
a few minor changes but overall this these are good updates

lirona committed:
custom errors

lirona commented:
added custom errors, note that if you [upgrade your solidity version ](https://soliditylang.org/blog/2024/09/04/solidity-0.8.27-release-announcement) you can use the nicer `require(bool, error)` pattern :)

daopunk commented on src/ButteredBread.sol:
```diff
@@ -123,25 +126,27 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
+        
+        _mint(_account, _amount * scalingFactors[_lp] / FIXED_POINT_PERCENT);
+        
+        bool success = IERC20(_lp).transferFrom(_account, address(this), _amount);
```
Why is transfer moved after mint? Funds should arrive to this contract, then tokens minted to the depositor.

lirona replied:
to follow the checks-effects-interactions pattern.
it's in the [commit message](https://github.com/BreadchainCoop/breadchain/pull/120/commits/c210616e12eeab396cb9ed15d3ab17cab125ab8c) :)

lirona replied:
to follow the checks-effects-interactions pattern, which reduces potential vulnerabilities like reentrancy.
and anyway it all happens in the same transactions so it either all happens or it all doesn't happen.. the order doesn't change the outcome.
actually to make the review process easier, i submitted issues for each fix and linked them to each commit that fixes them. this one fixes #116
welcome to look at the rest :)
#114 #115 #117 #118 #119

RonTuretzky replied:
Here is a nice blog posts on this [pattern](https://samvishwas.medium.com/solidity-programming-the-checks-effects-interactions-pattern-bac0e5a5f659#:~:text=The%20first%20step%20in%20the,proceed%20with%20the%20requested%20action.)

daopunk replied:
@RonTuretzky i wrote a article on this too few years ago 
https://medium.com/better-programming/solidity-smart-contract-security-preventing-reentrancy-attacks-fc729339a3ff

daopunk replied:
I get it is to follow the CEI pattern, but that usually applies to funds flowing out to the caller. in this case, funds are flowing into the contract, only then will coins be minted to the caller. so in this context, it seems minting first before receiving payment is more of a vulnerability, right?

RonTuretzky replied:
If I'm understanding correctly, what we're afraid of is a re-entrancy attack from the LP. 

If the re-entrancy entrypoint is in the `transferFrom` , it means the attack uses `transferFrom` to call `deposit` again

so this is the attack in the currently deployed implementation

```mermaid 
sequenceDiagram 
autonumber
Attacker ->> BB : deposit
loop  
BB ->> CorruptedLP : transferFrom
CorruptedLP ->> BB : deposit
end
BB ->> BB : mint
```

I believe the proposed change would make it so that this is possible instead 

```mermaid 
sequenceDiagram 
autonumber
Attacker ->> BB : deposit
loop  
BB ->> BB : mint
BB ->> CorruptedLP : transferFrom
CorruptedLP ->> BB : deposit
end
```

Which seems worst to me

daopunk replied:
Exactly.

Notice in the _withdraw function, the balance is updated before being sent to the caller, whereas in the _deposit function the balance is updated after the funds are received to the contract. It is the same pattern but applied in reverse depending on the context.

receive funds from caller => mint coins to caller
burn coins from caller => release funds to caller

RonTuretzky replied:
updated the diagram to reflect the mint before transferFrom

daopunk replied:
Might be worth it to add a re-entrancy guard tho @lirona

lirona replied:
great catch! sorry guys, didn't think this through.. good lesson for me that best practices are not always best :))
i will revert this change and add a reentrancy guard 🙏

lirona replied:
moved back in https://github.com/BreadchainCoop/breadchain/pull/120/commits/4fc9568be5c3ac64a0e28e082bf3ebf407e27761 and added reentrancy guard in https://github.com/BreadchainCoop/breadchain/pull/120/commits/73bc7fc447979812b2856913bea23733ef0a755b
didn't add on the `withdraw` function yet, on one hand can't hurt, but it will add some gas cost (to reduce the gas cost we can implement it using [`transient` which is supposed to reduce the gas cost significantly](https://docs.soliditylang.org/en/v0.8.28/contracts.html#transient-storage), but i'm guessing OZ will update their implementation to it soon)
wdyt?

RonTuretzky replied:
I guess we should add it to the withdraw flow as well, as there could be a re-entrancy as well over there if we're already covering it

lirona replied:
added, so resolving this :)

bagelface reviewed (approved):
lgtm

lirona committed:
add comment about ButteredBread changes

RonTuretzky reviewed (approved):
LGTM!

lirona committed:
revert move transferFrom

lirona committed:
add reentrancy guard on deposit

also added __ERC20Votes_init which does nothing but might change in future versions

lirona committed:
add nonReentrant on withdraw
